### PR TITLE
[Operator Versioning][Edge] Codegen upgrader_mobile.cpp

### LINKED ISSUE #69905: `torch.jit.operator_upgraders` introduces new runtime dependency to `yaml`
### 🐛 Describe the bug

`import yaml` was added to `torch/jit/operator_upgraders.py` by https://github.com/pytorch/pytorch/pull/68772
Which is not properly reflected as `setup.py` dependency. 

TODO: Revert https://github.com/pytorch/vision/commit/743fe1ab802c27c52a5372577d006d595e65677b after issue is closed

### Versions

https://download.pytorch.org/whl/nightly/cpu/torch-1.11.0.dev20211214%2Bcpu-cp36-cp36m-linux_x86_64.whl

cc @ezyang @gchanan @zou3519 @bdhirsh @malfet @seemethere

### PR DESCRIPTION
**Pull request by cccclai**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69194

From operator version map and upgrader torchscript, generate upgrader_mobile.cpp file. It also includes a unit test.

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D32748985/)!

Differential Revision: [D32748985](https://our.internmc.facebook.com/intern/diff/D32748985)

### CODE REVIEW

**Review by iseeyuan — COMMENTED** [review]
I think we only need to commit one generated file, either yaml or the header file. Looks like committing the header file makes more sense?

**Comment by cccclai** [comment]
> I think we only need to commit one generated file, either yaml or the header file. Looks like committing the header file makes more sense?

Currently no files are checked in. The checked in two files are for testing purposes at the moment. Likely Tugsuu will need to use this file to generate cpp.

**Review by iseeyuan — APPROVED** [review]
LGTM.

File: tools/codegen/operator_versions/gen_mobile_upgraders.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+        "--output_path",
+        help=f"Output {UPGRADER_MOBILE_FILE_NAME}.",
+        required=True,
+    )
+    options = parser.parse_args()
+    return options
+
+
+def main(argv):
+    parser = argparse.ArgumentParser(description="Generate used operators YAML")
```
Comment (gmagogsfm) [inline tools/codegen/operator_versions/gen_mobile_upgraders.py:272]: Description seems to be for a different argument?

File: tools/codegen/operator_versions/gen_mobile_upgraders.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+         std::vector<Upgrader>({Upgrader({0, 3, "div_Tensor_0_3", 0})})},
+        {std::string("aten::div.Scalar"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div_Scalar_0_3", 1})})},
+        {std::string("aten::div.out"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div_out_0_3", 2})})},
+        {std::string("aten::div_.Tensor"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div__Tensor_0_3", 3})})},
+        {std::string("aten::div_.Scalar"),
+         std::vector<Upgrader>({Upgrader({0, 3, "div__Scalar_0_3", 4})})},
+    });
```
Comment (gmagogsfm) [inline tools/codegen/operator_versions/gen_mobile_upgraders.py:107]: Why are upgrader definitions hard coded here?
@tugsbayasgalan I think you mentioned that operator_versions.yaml will be deleted, then what will replace it? It seems we are still referencing it here.

File: tools/codegen/operator_versions/gen_mobile_upgraders.py (hunk 1 above)
Comment (cccclai) [reply tools/codegen/operator_versions/gen_mobile_upgraders.py:344]: Oh good call. Update in the latest commit.

File: tools/codegen/operator_versions/gen_mobile_upgraders.py (hunk 2 above)
Comment (cccclai) [reply tools/codegen/operator_versions/gen_mobile_upgraders.py]: Yeah this python will take an input bytecode yaml and output cpp. 

In the latest commit, `kOperatorVersionMap` is generated from `torch._C._get_operator_version_map()`, which is populated from `caffe2/torch/csrc/jit/operator_upgraders/version_map.h`  in https://github.com/pytorch/pytorch/pull/68772.

Comment (cccclai) [reply tools/codegen/operator_versions/gen_mobile_upgraders.py]: In https://github.com/pytorch/pytorch/pull/68005, `def generate_bytecode(file_name)` will generate a bytecode yaml file, I will append this function at the end of `generate_bytecode` to generate a bytecode cpp instead.

Comment (cccclai) [reply tools/codegen/operator_versions/gen_mobile_upgraders.py]: The reason why codegen another version map is that it needs to map the the index in `std::vector<ByteCodeFunctionWithOperator> kUpgraderByteCode`. When replacing the operator in bytecode
`OP`, `X`, `N`, we can replace `OP` with `CALL` and `X` with the according index in `kUpgraderByteCode`. 

Maybe there is a more convenient way to replace. What I describe above is the implementation in https://github.com/pytorch/pytorch/pull/67731

File: test/mobile/test_upgrader_bytecode_table_example.h [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+struct Upgrader {
+  int min_version;
+  int max_version;
+  std::string upgrader_name;
+  int index;
+};
+
+// From operator_versions.yaml
+
+const std::unordered_map<std::string, std::vector<Upgrader>> kOperatorVersionMap(
```
Comment (tugsbayasgalan) [inline test/mobile/test_upgrader_bytecode_table_example.h:32]: Can this be a different name if we are using kOperatorVersionMap for the actual version_map.h

File: tools/codegen/operator_versions/gen_mobile_upgraders.py [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+
+def construct_version_maps(upgrader_bytecode_function_to_index_map: Dict) -> Dict:
+    version_map = torch._C._get_operator_version_map()
+    operator_list_in_version_map_part = []
+    for op_name in version_map:
+        upgraders_in_version_map_part = []
+        for upgrader_entry in version_map[op_name]:
+            # Split a string by "_" and filter empty string in the list
+            # For example: "div__Scalar_0_3" => ['div', 'Scalar', '0', '3']
+            upgrader_info = list(filter(lambda token: token != "", upgrader_entry.upgrader_name.split('_')))
```
Comment (tugsbayasgalan) [inline tools/codegen/operator_versions/gen_mobile_upgraders.py:233]: I think it is not safe to rely on the upgrader name to figure out the start and end. I think it rather should be based on version bump value. So for example, if the version bump happened at 4, we know that the range will be from 0 to 3.

File: test/mobile/test_upgrader_bytecode_table_example.h (hunk 3 above)
Comment (cccclai) [reply test/mobile/test_upgrader_bytecode_table_example.h:32]: Yeah I will rename it

**Pull request closed without merging**